Mathlib/Tactic/Hint.lean

```diff
@@ -0,0 +1,132 @@
+/-
+Copyright (c) 2023 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import Std.Tactic.TryThis
+import Std.Linter.UnreachableTactic
+import Mathlib.Data.Nondet.Basic
+import Mathlib.Tactic.FailIfNoProgress
+import Mathlib.Mathport.Rename
+
+#align_import tactic.hint from "leanprover-community/mathlib"@"8f6fd1b69096c6a587f745d354306c0d46396915"
+
+/-!
+# The `hint` tactic.
+
+The `hint` tactic tries the kitchen sink:
+it runs every tactic registered via the `register_hint tac` command
+on the current goal, and reports which ones succeed.
+
+## Future work
+It would be nice to run the tactics in parallel.
+-/
+
+open Lean Elab Tactic
+
+open Std.Tactic.TryThis
+
+namespace Mathlib.Tactic.Hint
+
+/-- An environment extension for registering hint tactics. -/
+initialize hintExtension : SimplePersistentEnvExtension (TSyntax `tactic) (List (TSyntax `tactic)) ←
+  registerSimplePersistentEnvExtension {
+    addEntryFn := (·.cons)
+    addImportedFn := mkStateFromImportedEntries (·.cons) {}
+  }
+
+/-- Register a new hint tactic. -/
+def addHint (stx : TSyntax `tactic) : CoreM Unit := do
+  modifyEnv fun env => hintExtension.addEntry env stx
+
+/-- Return the list of registered hint tactics. -/
+def getHints : CoreM (List (TSyntax `tactic)) := return hintExtension.getState (← getEnv)
+
+open Lean.Elab.Command in
+/--
+Register a tactic for use with the `hint` tactic, e.g. `register_hint simp_all`.
+-/
+elab (name := registerHintStx) "register_hint" tac:tactic : command => liftTermElabM do
+  -- remove comments
+  let tac : TSyntax `tactic := ⟨tac.raw.copyHeadTailInfoFrom .missing⟩
```

Comment:
I'm curious if this is the right thing to do. I know there's `tac.raw.unsetTrailing` as well, which doesn't clear source position information.

Maybe source position doesn't matter since it's not going to be used for any error reporting.

Reply:
I'm not sure.

If anything, removing all source position here seems safer: it would be terribly wrong to use it!